Anisotropic creep and plasticity

recuero opened the pull request:
This PR adds anisotropic (Hill-like) creep and plasticity to a hierarchy of generalized radial return classes in the `TensorMechanics` module.

Refs. #16016

recuero committed:
Initial commit for anisotropic plasticity Refs. #16016

recuero committed:
Pass tensors by reference

recuero committed:
Use DenseVector and DenseMatrix

recuero committed:
Move computeReferenceResidual to the material. More general

recuero committed:
Update inelastic strain tensor upon convergence of plastic multiplier (orthotropic creep model).

recuero committed:
Move to an AD hierarchy

recuero committed:
New AD hierarchy for anisotropic plasticity

recuero committed:
Toy problem for anisotropic plasticity 'running'

recuero committed:
Anisotropic creep model converges, but does not creep.

recuero committed:
Simple 2D model working (anis_mech_hill_tensor_creep_small.i). Next plasticity and verification.

recuero committed:
Create Hill creep model 2D tests

recuero committed:
Add anisotropic plasticity classes

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicPlasticityStressUpdate.C:
```diff
@@ -0,0 +1,378 @@
+  for (unsigned int i = 0; i < 6; i++)
+    inv_matrix(i, i) = 1 / (1 + _two_shear_modulus * delta_gamma * _eigenvalues_hill(i));
+
+  ADDenseVector stress_vector(6);
+  stress_vector(0) = stress_new(0);
+  stress_vector(1) = stress_new(1);
+  stress_vector(2) = stress_new(2);
+  stress_vector(3) = stress_new(3);
+  stress_vector(4) = stress_new(4);
+  stress_vector(5) = stress_new(5);
```
does `stress_vector = stress_new;` not work?

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicPlasticityStressUpdate.C:
```diff
@@ -0,0 +1,378 @@
+  //  Moose::out << "Print A matrix: " << MetaPhysicL::raw_value(A) << "\n";
+
+  Eigen::SelfAdjointEigenSolver<AnisotropyMatrix> es(A);
+  //  std::cout << "The eigenvalues of the pencil (A) are:" << std::endl
+  //            << es.eigenvalues() << std::endl;
+  //  std::cout << "The matrix of eigenvectors, V, is:" << std::endl
+  //            << es.eigenvectors() << std::endl
+  //            << std::endl;
```
remove all this stuff. it makes the pre-check fail!

recuero committed:
Add gold files, remove banned keyword

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/637632/) on ca2597b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16336/site)

This comment will be updated on new commits.

recuero committed:
Fix non-unity build

recuero commented:
@dschwen any idea why dual operators appear missing for non-unity build? `MooseTypes.h` and others are added directly or indirectly -I believe.

recuero committed:
Update anisotropy PR to new generic RadialReturn class

recuero committed:
Anisotropic plasticity based on generalized radial return mapping

recuero committed:
Problem converges in few iterations (doesn't modify stress or strain)

recuero committed:
New features:

 - Make material time step work through effective_inelastic_strain
 - Make call to computeStressFinalize to modify Cauchy stress due to plastic yield
 - Disable debugging Moose::out statements

recuero committed:
Add plastic tests that ensure:

 - Anisotropic plastic test behaves like isotropic elastic if no plastic excusion takes place
 - Evolution of plastic strain and convergence looks okay by simple inspection (verification ahead)

Avoid compiler warnings for unused variables.

recuero committed:
Demote sanity check tests to first order to prevent having insufficient AD vector size

recuero committed:
Add markdown documentation to anisotropic creep and plasticity

dschwen commented:
This should get wrapped up. I don't feel good about a large amount of code hovering around as an open PR.

recuero commented:
Just finished plasticity verification. Better to get it in with some meaningful tests, imo.

recuero committed:
Fix accumulation of isotropic hardening variable

recuero committed:
Add verification-based regression tests for isotropic elasticity

recuero committed:
Fix tests in parallel

recuero committed:
Some code cleanup, doxygen comments

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicPlasticityStressUpdate.C:
```diff
@@ -0,0 +1,353 @@
+
+  for (unsigned int index_i = 0; index_i < dimension; index_i++)
+    for (unsigned int index_j = 0; index_j < dimension; index_j++)
+      _eigenvectors_hill(index_i, index_j) = es.eigenvectors()(index_i, index_j);
```
```suggestion
      _eigenvectors_hill(index_i, index_j) =v(index_i, index_j);
```

recuero replied:
I didn't commit your suggestion from here because I was afraid it was going to have a formatting issue

recuero committed:
Address reviewer's comment

dschwen commented on modules/tensor_mechanics/include/materials/ADGeneralizedRadialReturnStressUpdate.h:
```diff
@@ -0,0 +1,162 @@
+#include "StressUpdateBase.h"
+#include "ADGeneralizedReturnMappingSolution.h"
+
+/**
```
can you move this to right above the class definition?

dschwen commented on modules/tensor_mechanics/include/materials/ADGeneralizedRadialReturnStressUpdate.h:
```diff
@@ -0,0 +1,162 @@
+#include <Eigen/Dense>
+#include "metaphysicl/dualnumber.h"
+
+namespace Eigen
```
can we move this into the implementation file? or into MooseTypes.h

recuero replied:
Every child class would need to do this if we move it to the implementation file. Is this so bad?

dschwen commented on modules/tensor_mechanics/include/materials/ADAnisotropicReturnPlasticityStressUpdateBase.h:
```diff
@@ -0,0 +1,62 @@
+                                 const ADMaterialProperty<RankTwoTensor> & /*strain_rate*/)
+  {
+    mooseError(
+        "The computation of strain energy rate density needs to be implemented by a child class");
```
any reason this is not a pure virtual method?

dschwen commented on modules/tensor_mechanics/include/materials/ADGeneralizedRadialReturnStressUpdate.h:
```diff
@@ -0,0 +1,162 @@
+                                     const ADDenseVector & /*stress_dev*/,
+                                     const ADReal & /*delta_gamma*/)
+  {
+    mooseError("computeStrainFinalize needs to be implemented by a child class.");
```
again, this should be pure virtual to make it a compile error and not a runtime error.

dschwen commented on modules/tensor_mechanics/include/materials/ADTransverselyIsotropicCreepStressUpdate.h:
```diff
@@ -0,0 +1,117 @@
+   * @param stress Cauchy stresss tensor
+   * @param stress_dev Deviatoric partt of the Cauchy stresss tensor
+   * @param delta_gamma Generalized radial return's plastic multiplier
+   *
```
rm line and below and empty lines between class definition and doxygen comments

dschwen commented on modules/tensor_mechanics/src/materials/ADGeneralizedRadialReturnStressUpdate.C:
```diff
@@ -0,0 +1,155 @@
+
+  elastic_strain_increment -= inelastic_strain_increment;
+
+  // stress_new = elasticity_tensor * (elastic_strain_increment + elastic_strain_old);
```
remove?

dschwen commented on modules/tensor_mechanics/src/materials/ADSingleVariableReturnMappingSolution.C:
```diff
@@ -184,6 +184,9 @@ ADSingleVariableReturnMappingSolution::internalSolve(const ADReal effective_tria
   {
+    //    Moose::out << "**_residual: " << MetaPhysicL::raw_value(_residual) << "\n";
+    //    Moose::out << "_iteration: " << MetaPhysicL::raw_value(_iteration) << "\n";
+    //    Moose::out << "reference_residual: " << reference_residual << "\n";
```
remove

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicCreepStressUpdate.C:
```diff
@@ -0,0 +1,247 @@
+               "Start time must be equal to or greater than the Executioner start_time if a "
+               "non-integer m_exponent is used");
+
+  _hill_constants = getParam<std::vector<Real>>("hill_constants");
```
this is never length-checked

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicCreepStressUpdate.C:
```diff
@@ -0,0 +1,247 @@
+  params.addRequiredParam<Real>("activation_energy", "Activation energy");
+  params.addParam<Real>("gas_constant", 8.3143, "Universal gas constant");
+  params.addParam<Real>("start_time", 0.0, "Start time (if not zero)");
+  params.addRequiredParam<std::vector<Real>>("hill_constants",
```
```
params.addRequiredRangeCheckedParam<std::vector<Real>>("hill_constants", "hill_constants_size = 6",
```

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicCreepStressUpdate.C:
```diff
@@ -0,0 +1,247 @@
+    const ADReal & delta_gamma)
+{
+  // Hill constants, some constraints apply
+  const Real F = _hill_constants[0];
+  const Real G = _hill_constants[1];
+  const Real H = _hill_constants[2];
+  const Real L = _hill_constants[3];
+  const Real M = _hill_constants[4];
+  const Real N = _hill_constants[5];
```
Semantically it seems better to use `const &` here as you are declaring an alias. That way the compiler can substitute the RHS in below without ever crating a copy. If you don't use a reference there is no guarantee that LHS and RHS are the same (`_hill_constants` could be changed as a side-effect, especially since it is not `const` itself). Way too long of a comment for such a minor point...

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicPlasticityStressUpdate.C:
```diff
@@ -0,0 +1,353 @@
+  const Real H = _hill_constants[2];
+  const Real L = _hill_constants[3];
+  const Real M = _hill_constants[4];
+  const Real N = _hill_constants[5];
```
see above (several instances)

dschwen commented on modules/tensor_mechanics/src/materials/ADTransverselyIsotropicPlasticityStressUpdate.C:
```diff
@@ -0,0 +1,353 @@
+ADTransverselyIsotropicPlasticityStressUpdate::computeHillTensorEigenDecomposition(
+    ADDenseMatrix & hill_tensor)
+{
+  unsigned int dimension = hill_tensor.n();
```
const

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_creep/anis_mech_hill_tensor_creep_small_tiny_step_ts_limit_test.i:
```diff
@@ -0,0 +1,247 @@
+    C_ijkl = '2.0e3 2.0e5 2.0e3 0.71428571e3 0.71428571e3 0.71428571e3 0.4 0.2 0.004 0.004 0.2 0.4'
+  []
+
+  #  [./elasticity_tensor]
+  #      type = ADComputeIsotropicElasticityTensor
+  #      youngs_modulus = 206800
+  #      poissons_ratio = 0.0
+  #    [../]
+
+  #  [stress_]
+  #     type = ADComputeFiniteStrainElasticStress
+  #  []
```
rm

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_creep/anis_mech_hill_tensor_creep_small_tiny_step_test.i:
```diff
@@ -0,0 +1,237 @@
+    []
+  []
+
+  #  [./pull_disp_y]
+  #    type = ADFunctionDirichletBC
+  #    variable = disp_y
+  #    boundary = top
+  #    function = pull
+  #  [../]
```
rm (all commented out parts)

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_plasticity/ad_aniso_plasticity_x.i:
```diff
@@ -0,0 +1,281 @@
+[Modules/TensorMechanics/Master]
+  [all]
+    strain = FINITE
+    add_variables = true
```
you have `add_variables = true` but also add the variables explicitly above. Check inputs and avoid doing that.

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_plasticity/ad_aniso_plasticity_x.i:
```diff
@@ -0,0 +1,281 @@
+
+[]
+
+[Preconditioning]
+  [smp]
+    type = SMP
+    full = true
+  []
+[]
+
```
Moose has `Executioner/auto_preconditioning = true` by default now. So for Newton this block is not necessary anymore.

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_plasticity/ad_aniso_plasticity_y.i:
```diff
@@ -0,0 +1,281 @@
+  [all]
+    strain = FINITE
+    add_variables = true
+    incremental = true
```
for finite strain `incremental` defaults to true...

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_plasticity/ad_aniso_plasticity_y.i:
```diff
@@ -0,0 +1,281 @@
+      function = pull
+    []
+  []
+
```
I'd recommend using new HIT syntax in all new files (and using `hit format` to prettify the inputs, e.g. to get rid of stray empty lines and ensure consistent indentation. there is an Atom plugin for that...)

recuero replied:
That empty line isn't removed by the Atom plugin

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_plasticity/tests:
```diff
@@ -0,0 +1,38 @@
+    input = 'anis_plasticity_test.i'
+    csvdiff = 'anis_plasticity_test_out.csv'
+    requirement = 'Anisotropic plasticity must yield same results as finite strain elasticity if yield condition is never met'
+  [../]
```
you can use new HIT syntax here, too

dschwen reviewed (commented):
Looks beautiful. A few small comments.

recuero committed:
Address reviewer's comments

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_plasticity/ad_aniso_plasticity_y.i:
```diff
@@ -0,0 +1,272 @@
+  []
+  [plasticity_strain_yy]
+    type = ADRankTwoAux
+    rank_two_tensor = trial_plasticity_plasticity_strain
+    variable = plasticity_strain_yy
```
we have `plastic_strain_??` as a standard material property. The two names here are a bit confusing.

dschwen replied:
and any reason you removed the one `add_variables` line rather than the variables block above?

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_plasticity/ad_aniso_plasticity_y.i:
```diff
@@ -0,0 +1,272 @@
+    order = CONSTANT
+    family = MONOMIAL
+  []
+  [stress_yy]
```
you add stress_yy here but not stress_xx. both are added via `generate_outputs` in the master action

recuero committed:
Move definition of eigendecomposition to MooseTypes, simplify input files

dschwen commented:
Looks like while moving teh eigen stuff to moosetypes you dropped the `#include < Eigen/Eigenvalues>`

recuero commented:
I dropped that header because including it was causing a `libmesh` error -- not including it causes non unity build to fail. This is the error

```Compiling C++ (in opt mode) /opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C...
cc1plus: all warnings being treated as errors
make: *** [/opt/civet/build_0/moose/framework/build.mk:118: /opt/civet/build_0/moose/modules/tensor_mechanics/build/unity_src/utils_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
In file included from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/libmesh/include/Eigen/Geometry:40,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/libmesh/include/Eigen/Eigenvalues:12,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/MooseTypes.h:39,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/InputParameters.h:15,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/DistributionInterface.h:12,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/UserObject.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/ElementUserObject.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/ShapeElementUserObject.h:12,
                 from /opt/civet/build_0/moose/modules/phase_field/build/header_symlinks/ComputeExternalGrainForceAndTorque.h:12,
                 from /opt/civet/build_0/moose/modules/phase_field/src/userobjects/ComputeExternalGrainForceAndTorque.C:10,
                 from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/userobjects_Unity.C:2:
/opt/civet/build_0/conda_builds/conda_envs/pr-16336/libmesh/include/Eigen/src/Geometry/Quaternion.h: In instantiation of 'void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = Eigen::Quaternion<double>; _Args = {const Eigen::Quaternion<double, 0>&}; _Tp = Eigen::Quaternion<double>]':
/opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/bits/alloc_traits.h:484:4:   required from 'static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = Eigen::Quaternion<double>; _Args = {const Eigen::Quaternion<double, 0>&}; _Tp = Eigen::Quaternion<double>; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<Eigen::Quaternion<double> >]'
/opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/bits/stl_vector.h:1189:30:   required from 'void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = Eigen::Quaternion<double>; _Alloc = std::allocator<Eigen::Quaternion<double> >; std::vector<_Tp, _Alloc>::value_type = Eigen::Quaternion<double>]'
/opt/civet/build_0/moose/modules/phase_field/src/userobjects/EBSDReader.C:226:52:   required from here
/opt/civet/build_0/conda_builds/conda_envs/pr-16336/libmesh/include/Eigen/src/Geometry/Quaternion.h:226:7: error: implicitly-declared 'constexpr Eigen::QuaternionBase<Eigen::Quaternion<double> >::QuaternionBase(const Eigen::QuaternionBase<Eigen::Quaternion<double> >&)' is deprecated [-Werror=deprecated-copy]
  226 | class Quaternion : public QuaternionBase<Quaternion<_Scalar,_Options> >
      |       ^~~~~~~~~~
/opt/civet/build_0/conda_builds/conda_envs/pr-16336/libmesh/include/Eigen/src/Geometry/Quaternion.h:477:46: note: because 'Eigen::QuaternionBase<Eigen::Quaternion<double> >' has user-provided 'Eigen::QuaternionBase<Derived>& Eigen::QuaternionBase<Derived>::operator=(const Eigen::QuaternionBase<Derived>&) [with Derived = Eigen::Quaternion<double>]'
  477 | EIGEN_STRONG_INLINE QuaternionBase<Derived>& QuaternionBase<Derived>::operator=(const QuaternionBase<Derived>& other)
      |                                              ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/x86_64-conda-linux-gnu/bits/c++allocator.h:33,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/bits/allocator.h:46,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/string:41,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/bits/locale_classes.h:40,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/bits/ios_base.h:41,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/ios:42,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/istream:38,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/sstream:38,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/complex:45,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/libmesh/include/libmesh/libmesh_common.h:46,
                 from /opt/civet/build_0/conda_builds/conda_envs/pr-16336/libmesh/include/libmesh/threads.h:24,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/HashMap.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/MooseUtils.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/InputParameters.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/DistributionInterface.h:12,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/UserObject.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/ElementUserObject.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/ShapeElementUserObject.h:12,
                 from /opt/civet/build_0/moose/modules/phase_field/build/header_symlinks/ComputeExternalGrainForceAndTorque.h:12,
                 from /opt/civet/build_0/moose/modules/phase_field/src/userobjects/ComputeExternalGrainForceAndTorque.C:10,
                 from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/userobjects_Unity.C:2:
/opt/civet/build_0/conda_builds/conda_envs/pr-16336/x86_64-conda-linux-gnu/include/c++/9.3.0/ext/new_allocator.h:147:4: note: synthesized method 'Eigen::Quaternion<double>::Quaternion(const Eigen::Quaternion<double>&)' first required here
  147 |  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make: *** [/opt/civet/build_0/moose/framework/build.mk:118: /opt/civet/build_0/moose/modules/phase_field/build/unity_src/userobjects_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
ERROR: exiting with code 2

recuero committed:
Clean up existing hierarchy

recuero committed:
Initial implementation/verification of elastoplastic anisotropy

recuero committed:
Change class names to 'Hill'

recuero committed:
Fix creep plastic multiplier bug and update tests

recuero committed:
Add Hill creep tests from analytical verification exercise

moosebuild commented:
Job [End User Experience (Ubuntu)](https://civet.inl.gov/job/683788/) on 1140096 : invalidated by @milljm

recuero committed:
Avoid Eigen's ambiguous operator error.

- Eigen matrices will be non-AD. For now, they can be considered as user input.
- Some clean up. Apply hit format to some inputs.

dschwen commented:
Just FYI, I created a small test repo (https://github.com/dschwen/eigen_metaphysicl_test) and commented on the corresponding Eigen bug (https://gitlab.com/libeigen/eigen/-/issues/1902).

I also pinged Roy on Slack on this issue. It would be nice to get to the bottom of it.

recuero committed:
Make anisotropic creep truly implicit
 - Before this commit, elastic strain was not fully correct
 - Update tests, successful comparison with isotropic power law
 - Address reviewer's comment on the use of "zero()" method

Anisotropic creep is sensitive to time step and stress increments. To be investigated.

recuero commented:
@dschwen please take a look. This seems like a good point to try to merge what we have.

dschwen commented on modules/tensor_mechanics/src/materials/ADHillCreepStressUpdate.C:
```diff
@@ -0,0 +1,264 @@
+  qsigma_square = std::sqrt(qsigma_square);
+  if (qsigma_square == 0)
+  {
+    inelasticStrainIncrement(0, 0) = inelasticStrainIncrement(1, 1) =
```
Why not `inelasticStrainIncrement.zero()`?

dschwen reviewed (dismissed):
Looks great

dschwen reviewed (approved):
👍

gambka commented:
It looks like this has finally made it into the moose submodule used in downstream apps and is causing debug testing to fail on those apps that have the tensor_mechanics module in them.
